near-sdk/src/json_types/public_key.rs

```diff
@@ -97,31 +128,38 @@ impl<'de> serde::Deserialize<'de> for Base58PublicKey {
     where
         D: serde::Deserializer<'de>,
     {
-        let s: String = Deserialize::deserialize(deserializer)?;
+        let s: String = serde::Deserialize::deserialize(deserializer)?;
         s.parse::<Base58PublicKey>().map_err(serde::de::Error::custom)
     }
 }
 
 impl From<&Base58PublicKey> for String {
     fn from(str_public_key: &Base58PublicKey) -> Self {
-        match str_public_key.0[0] {
-            0 => "ed25519:".to_string() + &bs58::encode(&str_public_key.0[1..]).into_string(),
-            1 => "secp256k1:".to_string() + &bs58::encode(&str_public_key.0[1..]).into_string(),
-            _ => panic!("Unexpected curve"),
-        }
+        let upper_bound = (str_public_key.data.len() / 5 + 1) * 8;
+        let (pad, curve_as_str) = match str_public_key.curve {
+            CurveType::ED25519 => (8, "ed25519:"),
+            CurveType::SECP256K1 => (10, "secp256k1:"),
+        };
+
+        // Efficient concat of "curve_type:" + bs58 encoded bytes:
+        let mut buf = vec![0u8; pad + upper_bound];
+        buf.splice(..pad, curve_as_str.as_bytes().iter().cloned());
+        let size = bs58::encode(&str_public_key.data).into(&mut buf[pad..]).unwrap();
+        buf.truncate(pad + size);
+        String::from_utf8(buf).unwrap()
```

Comment:
why not just encode the bs58 to a `&mut str` buffer? Can do `buf.get_mut(pad..)` if it's a string buffer and can do the same as above. The unwrap might bring in code not needed

Reply:
The thing is, `get_mut` returns an optional which is none when it's out of bounds. I don't know if there's an easy way to create a buffer string similar to `vec![0u8; size]` where we fill the whole capacity at the start.